ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java

```diff
@@ -78,203 +82,667 @@
+            replicasToReplicate.get(remoteHost.dataNodeId));
+    assertEquals("Response should contain a response for each replica",
+        replicasToReplicate.get(remoteHost.dataNodeId).size(), response.size());
+    for (int i = 0; i < response.size(); i++) {
+      if (i % 2 == 0) {
+        assertEquals(0, response.get(i).missingStoreKeys.size());
+        assertEquals(expectedIndex, ((MockFindToken) response.get(i).remoteToken).getIndex());
+      } else {
+        assertEquals(1, response.get(i).missingStoreKeys.size());
+        assertEquals(expectedIndex + 1, ((MockFindToken) response.get(i).remoteToken).getIndex());
+      }
+    }
+    replicaThread.fixMissingStoreKeys(new MockConnection(remoteHost, batchSize),
+        replicasToReplicate.get(remoteHost.dataNodeId), response);
+    for (int i = 0; i < response.size(); i++) {
+      assertEquals("Token should have been set correctly in fixMissingStoreKeys()", response.get(i).remoteToken,
+          replicasToReplicate.get(remoteHost.dataNodeId).get(i).getToken());
     }
 
-    @Override
-    public int count() {
-      return bytebuffers.size();
+    // no more missing keys
+    response = replicaThread.exchangeMetadata(new MockConnection(remoteHost, 4),
+        replicasToReplicate.get(remoteHost.dataNodeId));
+    assertEquals("Response should contain a response for each replica",
+        replicasToReplicate.get(remoteHost.dataNodeId).size(), response.size());
+    for (int i = 0; i < response.size(); i++) {
+      assertEquals(0, response.get(i).missingStoreKeys.size());
+      assertEquals(i % 2 == 0 ? expectedIndex : expectedIndex + 1,
+          ((MockFindToken) response.get(i).remoteToken).getIndex());
     }
 
-    @Override
-    public long sizeInBytes(int index) {
-      return bytebuffers.get(0).limit();
+    Map<PartitionId, List<MessageInfo>> missingInfos = remoteHost.getMissingInfos(localHost.infosByPartition);
+    for (Map.Entry<PartitionId, List<MessageInfo>> entry : missingInfos.entrySet()) {
+      // test that the first key has been marked deleted
+      List<MessageInfo> messageInfos = localHost.infosByPartition.get(entry.getKey());
+      StoreKey deletedId = messageInfos.get(0).getStoreKey();
+      assertNotNull(deletedId + " should have been deleted", getDeleteInfo(deletedId, messageInfos));
+      Map<StoreKey, Boolean> ignoreState = new HashMap<>();
+      for (StoreKey toBeIgnored : idsToBeIgnoredByPartition.get(entry.getKey())) {
+        ignoreState.put(toBeIgnored, false);
+      }
+      for (MessageInfo messageInfo : entry.getValue()) {
+        StoreKey id = messageInfo.getStoreKey();
+        if (!messageInfo.getStoreKey().equals(deletedId)) {
+          assertTrue("Message should be eligible to be ignored", ignoreState.containsKey(id));
+          ignoreState.put(id, true);
+        }
+      }
+      for (Map.Entry<StoreKey, Boolean> stateInfo : ignoreState.entrySet()) {
```

Comment:
[This|https://github.com/linkedin/ambry/pull/702/files/9fb3dfeb0a504512e1f1551f3a27e2f0b2ac1436#diff-a6592d17a231c8fc632c42511bcbdea8R262] is the call that compares it. 

The function implementation is [here|https://github.com/linkedin/ambry/pull/702/files/9fb3dfeb0a504512e1f1551f3a27e2f0b2ac1436#diff-a6592d17a231c8fc632c42511bcbdea8R527]